### PR TITLE
Support dynamic creds with non-renewable leases

### LINKED ISSUE #319: AWS Dynamic Credentials Not Rotated
**Describe the bug**

I have a VaultDynamicSecret which is issuing AWS credentials. On creation, VSO successfully fetches creds and creates the secret. After the credential TTL the creds are never refetched and are stuck with the initial value.

Restarting the VSO pods causes them to see the creds are expired (presumably using the .status.lastRenewalTime and .status.secretLease.leaseDuration fields) and correctly rotates them. But only once. And then it fails to rotate them after that.

**To Reproduce**

This is the VaultDynamicSecret I am deploying (I've included the status fields for reference, but obviously those are not in the actual deployed manifest):

```yaml
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  name: aws-creds
spec:
  destination:
    create: true
    name: aws-creds
  mount: aws
  path: creds/k8s-image-swapper
  renewalPercent: 75
  rolloutRestartTargets:
  - kind: Deployment
    name: k8s-image-swapper
  vaultAuthRef: default
status:
  lastGeneration: 1
  lastRenewalTime: 1691593888
  lastRuntimePodUID: c36f809c-446b-4d00-b089-f2d6990959b1
  secretLease:
    duration: 899
    id: aws/credsk8s-image-swapper/20lKjhw0e6MV7zMUldTQaKuP
    renewable: false
    requestID: cf67cd4a-fea4-bc99-caf0-8f408ec7baf8
  staticCredsMetaData:
    lastVaultRotation: 0
    rotationPeriod: 0
    ttl: 0
```

When the VSO pod is restarted it gives me this log message:

```
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T14:59:52Z	DEBUG	events	Not in renewal window after transitioning to a new leader/pod, lease_id=aws/creds/k8s-image-swapper/qaLmnOn90F2btwYK4WlI10sr, horizon=11m34.715898713s	{"type": "Normal", "object": {"kind":"VaultDynamicSecret","namespace":"k8s-image-swapper","name":"aws-creds","uid":"667f4fcc-1713-434c-b3dd-f7b17d9a37e9","apiVersion":"secrets.hashicorp.com/v1beta1","resourceVersion":"67751556"}, "reason": "SecretLeaseRenewal"}
```

But then, after the horizon time it rotates the creds and gives a horizon of 0. (Notice how it _does_ correctly rotate the creds 1 time and restart the deployment as expected.)

```
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T15:11:28Z	DEBUG	events	Lease is not renewable, staticCreds=false, info=&v1beta1.VaultSecretLease{ID:"aws/creds/k8s-image-swapper/20lKjhw0e6MV7zMUldTQaKuP", LeaseDuration:899, Renewable:false, RequestID:"cf67cd4a-fea4-bc99-caf0-8f408ec7baf8"}	{"type": "Warning", "object": {"kind":"VaultDynamicSecret","namespace":"k8s-image-swapper","name":"aws-creds","uid":"667f4fcc-1713-434c-b3dd-f7b17d9a37e9","apiVersion":"secrets.hashicorp.com/v1beta1","resourceVersion":"67755382"}, "reason": "SecretLeaseRenewal"}
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T15:11:28Z	INFO	Vault secret does not support periodic renewal/refresh via reconciliation	{"controller": "vaultdynamicsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultDynamicSecret", "VaultDynamicSecret": {"name":"aws-creds","namespace":"k8s-image-swapper"}, "namespace": "k8s-image-swapper", "name": "aws-creds", "reconcileID": "94c14dfb-e03d-47cd-88e3-9a07339a6c83", "requeue": false, "horizon": "0s"}
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T15:11:28Z	DEBUG	events	Rollout restart triggered for {Deployment k8s-image-swapper}	{"type": "Normal", "object": {"kind":"VaultDynamicSecret","namespace":"k8s-image-swapper","name":"aws-creds","uid":"667f4fcc-1713-434c-b3dd-f7b17d9a37e9","apiVersion":"secrets.hashicorp.com/v1beta1","resourceVersion":"67755382"}, "reason": "RolloutRestartTriggered"}
```

This is my `sys/mounts/aws/tune`:

```json
{
  "default_lease_ttl": 900,
  "description": "",
  "force_no_cache": false,
  "max_lease_ttl": 43200
}
```

This is my `aws/roles/k8s-image-swapper`:

```json
{
  "credential_type": "assumed_role",
  "default_sts_ttl": 900,
  "iam_groups": [],
  "iam_tags": null,
  "max_sts_ttl": 900,
  "permissions_boundary_arn": "",
  "policy_arns": [],
  "policy_document": "",
  "role_arns": [
    "arn:aws:iam::12345:role/k8s-image-swapper"
  ],
  "user_path": ""
}
```

**Expected behavior**

The credentials should rotate every `renewalPercent * leaseDuration` or 675 seconds

**Environment**
* Kubernetes version: 
  * Distribution or cloud vendor : eks 1.24
* vault-secrets-operator version: 0.1.0

### PR DESCRIPTION
**Pull request by oliverisaac**
Fixes #319

### CODE REVIEW

**Comment by rovangju** [comment]
bump

**Comment by duong-se** [comment]
bump

File: controllers/vaultdynamicsecret_controller.go [hunk 1]
```diff
@@ -229,6 +229,13 @@ func (r *VaultDynamicSecretReconciler) Reconcile(ctx context.Context, req ctrl.R
 				"Secret synced, isStaticCreds=%t, horizon=%s, ttl=%d",
 				true, horizon, o.Status.StaticCredsMetaData.TTL)
 		}
+	} else {
+		// Not all dynamic credentials are renewable (e.g.: aws/creds/)
+		// In the scenario we resync the secret within the lease duration window
+		nextRenewalTime := nextRenewalTime(o)
+		horizon = nextRenewalTime.Sub(time.Now())
```
Comment (benashz) [inline controllers/vaultdynamicsecret_controller.go:236]: It seems possible that `horizon` could be less than or equal to 0 seconds, in the case where `nextRenewalTime()` is in the past. We will want to handle that case to ensure that we keep syncing, as is done here https://github.com/hashicorp/vault-secrets-operator/pull/320/files/17552b646993eaf812add00d6759ac0a0febb786#diff-162ee5e0791fe53245a14c3a3eb3ddc1e296cad32b7081a91cecc0becdeb9b0dR223

**Review by benashz — COMMENTED** [review]
Looking good! I've completed my initial review..

**Comment by benashz** [comment]
Hi @oliverisaac, thanks again for this PR, it solves a valid use-case. It would be great to get it in for the next VSO release, we'd be happy to pick it up if you think you won't have the spare cycles for it.

Ben

**Comment by benashz** [comment]
Closing in favor of #397. Thank you for your contribution @oliverisaac

**Pull request closed without merging**